### PR TITLE
[FIX] website_mail: recaptcha check was missing on contact creation

### PR DESCRIPTION
**Pull request by alessandro-caldonazzi**
The route `/website_mail/follow` can create new contacts, for that reason, it should be protected with recaptcha similar to `/website/form/<string:model_name>` and `/website_mass_mailing/subscribe`

This PR adds the Recaptcha check on the endpoint: `/website_mail/follow`

Before this PR it was possible to create as much dummy contact as you wanted as an attacker with a simple script

opw-4023724

### CODE REVIEW

File: addons/website_mail/static/src/js/follow.js [hunk 1]
```diff
@@ -2,11 +2,25 @@ odoo.define('website_mail.follow', function (require) {
...
+    const { ReCaptcha } = require("google_recaptcha.ReCaptchaV3");
 
 publicWidget.registry.follow = publicWidget.Widget.extend({
     selector: '#wrapwrap:has(.js_follow)',
     disabledInEditableMode: false,
 
+    /**
+     * @constructor
+     */
+    init: function() {
```
Comment (agau-odoo) [inline addons/website_mail/static/src/js/follow.js:14]:
```suggestion
    init() {
```

File: addons/website_mail/static/src/js/follow.js [hunk 2]
```diff
@@ -2,11 +2,25 @@ odoo.define('website_mail.follow', function (require) {
...
+        this._super(...arguments);
+        this._recaptcha = new ReCaptcha();
+    },
+
+    willStart: async function() {
```
Comment (agau-odoo) [inline addons/website_mail/static/src/js/follow.js:19]:
```suggestion
    async willStart() {
```

File: addons/website_mail/static/src/js/follow.js [hunk 3]
```diff
@@ -99,7 +113,7 @@ publicWidget.registry.follow = publicWidget.Widget.extend({
      * @private
      * @param {Event} ev
      */
-    _onClick: function (ev) {
+    _onClick: async function(ev) {
```
Comment (agau-odoo) [inline addons/website_mail/static/src/js/follow.js:116]:
```suggestion
    async _onClick(ev) {
```

File: addons/website_mail/controllers/main.py [hunk 4]
```diff
@@ -22,6 +24,17 @@ def website_message_subscribe(self, id=0, object=None, message_is_follower="on",
...
+            try:
+                if not request.env['ir.http']._verify_request_recaptcha_token('website_mail'):
+                    error_captcha = _("Suspicious activity detected by Google reCaptcha.")
+
+            except (ValidationError, UserError) as e:
+                error_captcha = e.args[0]
+            if error_captcha:
+                return json.dumps({
+                    'error': error_captcha,
+                })
```
Comment (agau-odoo) [inline addons/website_mail/controllers/main.py:37]: Why not verify recaptcha at the very beginning of the method?
See [this other json route you mentioned](https://github.com/odoo/odoo/blob/8873fe7240e505553650fc15aea1ea3a299acc0f/addons/website_mass_mailing/controllers/main.py#L28-L32)

File: addons/website_mail/static/src/js/follow.js [hunk 5]
```diff
@@ -111,14 +125,19 @@ publicWidget.registry.follow = publicWidget.Widget.extend({
         $jsFollow.removeClass('o_has_error').find('.form-control, .custom-select').removeClass('is-invalid');
 
         var email = $email.length ? $email.val() : false;
-        if (email || this.isUser) {
+        if ((email || this.isUser) && this._recaptchaLoaded) {
+            const tokenCaptcha = await this._recaptcha.getToken("website_mail");
+            if (!tokenCaptcha.token) {
+                return;
+            }
```
Comment (agau-odoo) [inline addons/website_mail/static/src/js/follow.js:132]: Should we fail silently?
[Here](https://github.com/odoo/odoo/blob/57793ff912ac5aab8f3d3e014992965d06290bf0/addons/website_mass_mailing/static/src/js/website_mass_mailing.js#L99-L108), it displays a notification.

I tested this locally, if no recaptcha key is configured then clicking on the button does nothing.
Do we want to fail in this case too?

**Review by agau-odoo — CHANGES_REQUESTED** [review]

File: addons/website_mail/static/src/js/follow.js (hunk 5 above)
Comment (alessandro-caldonazzi) [reply addons/website_mail/static/src/js/follow.js]: Indeed failing silently does not make sense, I just fixed that.

**Review by agau-odoo — CHANGES_REQUESTED** [review]
Small change request, otherwise looks very clean!

**Review by agau-odoo — APPROVED** [review]

**Review by bso-odoo — COMMENTED** [review]
Here are a few comments.

Did you make sure it also works when recaptcha is not setup ?

**Comment by alessandro-caldonazzi** [comment]
@bso-odoo 
> Did you make sure it also works when recaptcha is not setup ?

Yes, it works even if no captcha keys are set

**Comment by bso-odoo** [comment]
@robodoo r+

**Pull request closed without merging**